src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+     {!Sys.int_size}.
+  *)
+  let pint_raw (st : RS.t) : int =
+    if Sys.word_size = 32
+    then RS.bits st
+    else (* word size = 64 *)
+      (* Bottom 30 bits *)
+      let right = RS.bits st in
+      (* Middle 30 bits *)
+      let middle = (RS.bits st lsl 30) in
+      (* Technically we could write [3] but this is clearer *)
+      let two_bits_mask = 0b11 in
+      (* Top 2 bits *)
+      let left = ((RS.bits st land two_bits_mask) lsl 60) in
+      left lor middle lor right
+
+  let pint ?(origin : int = 0) : int t = fun st ->
+    let origin = parse_origin "Gen.pint" Format.pp_print_int ~origin ~low:0 ~high:max_int in
+    let x = pint_raw st in
+    Tree.make_primitive (Shrink.int_towards origin) x
+
+  let number_towards = Shrink.number_towards
+
+  let int_towards = Shrink.int_towards
+
+  let int64_towards = Shrink.int64_towards
+
+  let int32_towards = Shrink.int32_towards
+
+  let float_towards = Shrink.float_towards
+
+  let int : int t =
+    bool >>= fun b ->
+    if b
+    then pint ~origin:0 >|= (fun n -> - n - 1)
+    else pint ~origin:0
+
+  let int_bound (n : int) : int t =
+    if n < 0 then invalid_arg "Gen.int_bound";
+    fun st ->
+      if n <= (1 lsl 30) - 2
+      then Tree.make_primitive (Shrink.int_towards 0) (Random.State.int st (n + 1))
+      else Tree.map (fun r -> r mod (n + 1)) (pint st)
+
+  (** Shrink towards [origin] if provided, otherwise towards the middle of the range
+      (e.g. [int_range (-5) 15] will shrink towards [5])
+
+      To support ranges wider than [Int.max_int], the general idea is to find the center,
+      and generate a random half-difference number as well as whether we add or
+      subtract that number from the center. *)
+  let int_range ?(origin : int option) (a : int) (b : int) : int t =
+    if b < a then invalid_arg "Gen.int_range: high < low";
+    let origin = Option.fold ~none:a ~some:(fun origin ->
+        if origin < a
+        then invalid_arg "Gen.int_range: origin < low"
+        else if origin > b then invalid_arg "Gen.int_range: origin > high"
+        else origin) origin in
+    fun st ->
+      let Tree.Tree(n, _shrinks) = if a >= 0 || b < 0 then (
+          (* range smaller than max_int *)
+          assert (b-a >= 0);
+          Tree.map (fun n -> a + n) (int_bound (b - a) st)
+        ) else (
+          (* range potentially bigger than max_int: we split on 0 and
+             choose the itv wrt to their size ratio *)
+          let f_a = float_of_int a in
+          let ratio = (-.f_a) /. (1. +. float_of_int b -. f_a) in
+          if Random.State.float st 1. <= ratio
+          then Tree.map (fun n -> - n - 1) (int_bound (- (a+1)) st)
+          else int_bound b st
+        ) in
+      Tree.make_primitive (Shrink.int_towards origin) n
+
+  let (--) low high = int_range ~origin:low low high
+
+  let oneof (l : 'a t list) : 'a t =
+    int_range 0 (List.length l - 1) >>= List.nth l
```

Comment:
I confess I took inspiration for the "Shrink towards the first of the list" behaviour from Hedgehog: https://hackage.haskell.org/package/hedgehog-1.0.5/docs/Hedgehog-Gen.html#v:choice

While this is in no way a prerequisite (I think?), I would find it a bit weird that it doesn't shrink towards the same thing depending on the starting point (generated value). 

So I am rather in favor of keeping this behavior.

What I like about it is that when I write generators for sum types, I put what I consider (functionally speaking) the smallest variants first, so that shrinking will jump to "smaller" variants.
I guess what annoys you is that when no variant is "smaller" than the others, then this feels "arbitrary" to shrink towards the first?

Of course, there is also the possibility of creating 2 functions (or 1 with an optional parameter):
* shrink towards the first
* only shrink on the current "branch", don't jump to another generator

Reply:
I can't give a proper answer but in my "very small knowledge", I think that shrinking using bind can be annoying. 
Maybe @nrinaudo can give more feedback about that?

Reply:
I can't give a proper answer but in my "very small knowledge", I think that shrinking using bind can be annoying. 
Maybe @nrinaudo can give more feedback about that?

Reply:
I don't actually think I have much to contribute to this discussion, especially since I don't know much about OCaml and its idioms (or what this PR is trying to achieve...)

From a general perspective, there's no inherent problem with using bind to compose generators - it's just sometimes suboptimal. The problem is using it when you have independent generators, as it'll end up leaving entire branches of the shrink tree unexplored.

In this specific scenario, it fees like there's a dependency - the generator used depends on the generator chosen...

The problem is interesting, however, in that it exposes an important design choice: do you want shrinkers to find the smallest possible failing test case, or the smallest possible failing test case *that triggers the same flaw as the original one*?

For what it's worth, I have the impression most people want the later but settle for the former. There *are* a few oddballs that embrace the former wholeheartedly, like Hypothesis: it doesn't even look at test cases when shrinking, but will shrink the underlying sequence of random numbers (:exploding_head: ).

Reply:
> I'm not sure `>>=` has the proper shrinking behavior here (do we _want_ to shrink towards the beginning of the list?). 

Erlang QuickCheck (like Hedgehog) also shrinks a selector towards the beginning, see, e.g., https://robertoaloi.github.io/erlang/notes-on-erlang-quickcheck

I agree with the advantages that @Sir4ur0n point out in favor of keeping this behavior.

Reply:
I do have use for a shrink-towards-the-beginning of the list kind of behaviour. In fact I wrote one such generator this week (though, to be fair, it may have been a case of me needing a type-based shrinker, but it's good not to have to think about it).

That being said, it's probably possible to write a selective version of such a combinator

```ocaml
choice : 'a t list -> 'a t
choice = function
  [] -> fail_with "at least one"
  [x] -> x
  x::l -> ifS bool x (choice l)
```

You need to correct for bias in the coin-flipping, but I don't want to drag this example too much.

The one issue with this is that this can be a very inefficient shrinker with the definition I proposed elsewhere: the shrinker, I believe, shrinks in all of the generators, then ignores those shrinks which are not relevant with the currently chosen boolean.

There is also the question of where the boolean choice shrinks to? If it shrinks towards `false`, then we have strictly more shrink that in the current `choice` definition.

I wouldn't worry too much about this, though, as this can be fiddled with in future iterations. The shrinking behaviour can change without breaking people's code.